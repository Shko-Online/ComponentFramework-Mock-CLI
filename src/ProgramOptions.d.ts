import { OptionValues } from "commander";

export interface ProgramOptions extends OptionValues {
  component: string;
  outputFolder: string;
  renderGenerator: string;
  logLevel: "none" | "error" | "verbose";

}
