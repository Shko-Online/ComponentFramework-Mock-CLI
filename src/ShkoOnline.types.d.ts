export type IEnumValue = {
  name: string;
  value: string;
};

export type IProperty = {
  defaultValue?: string;
  name: string;
  required?: boolean;
  usage: "input" | "bound" | "output";
} & (
  | {
      type:
        | "Currency"
        | "DateAndTime.DateAndTime"
        | "DateAndTime.DateOnly"
        | "Decimal"
        | "FP"
        | "Lookup.Simple"
        | "MultiSelectOptionSet"
        | "Multiple"
        | "Object"
        | "OptionSet"
        | "SingleLine.Email"
        | "SingleLine.Phone"
        | "SingleLine.Text"
        | "SingleLine.TextArea"
        | "SingleLine.Ticker"
        | "SingleLine.URL"
        | "TwoOptions"
        | "Whole.None";
    }
  | {
      type: "Enum";
      enumValues: IEnumValue[];
    }
);

export type IData = {
  /**
   * Class Name of the component
   */
  Component: string;
  /**
   * Component folder relative to the renderGenerator
   */
  ComponentFolder: string;
  /**
   * Component index file relative to the ComponentFolder. Defined in ControlManifest.Input.xml in the resources node like: `<code path="index.ts" order="1" />`
   */
  ComponentIndex: string;
  /**
   * List of inputs
   */
  Inputs: IProperty[];

  /**
   * Mock Generator type (Starndard | React)
   */
  MockGenerator:
    | "ComponentFrameworkMockGenerator"
    | "ComponentFrameworkMockGeneratorReact";
};
