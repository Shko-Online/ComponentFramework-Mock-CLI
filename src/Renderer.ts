import { IData } from 'ShkoOnline.types';
import { readFile } from 'fs/promises';
import Handlebars from 'handlebars';

export default class Renderer {
    constructor() {
        Handlebars.registerHelper('GetPropertyMocks', this.GetPropertyMocks.bind(this));
        Handlebars.registerHelper('ControlToPropertyMock', this.ControlToPropertyMock.bind(this));
        Handlebars.registerHelper('ControlToType', this.ControlToType.bind(this));
        Handlebars.registerHelper('isVirtualComponent', this.isVirtualComponent.bind(this));
    }

    private GetPropertyMocks(inputs: IData['Inputs']) {
        const propertyMocks: string[] = [];
        Object.getOwnPropertyNames(inputs).forEach(p => {
            if (!propertyMocks.some(pm => pm === inputs[p])) {
                propertyMocks.push(this.ControlToPropertyMock(inputs[p]));
            }
        });
        return propertyMocks.join(', ');
    }

    private isVirtualComponent(mockGenerator: IData['MockGenerator']) {
        return mockGenerator === 'ComponentFrameworkMockGeneratorReact';
    }

    private ControlToPropertyMock(control: IData['Inputs'][any]) {
        switch (control) {
            case 'DateAndTime.DateOnly':
                return 'DateTimePropertyMock';
            case 'Multiple':
                return 'StringPropertyMock';
            case 'SingleLine.Text':
                return 'StringPropertyMock';
            case 'TwoOptions':
                return 'TwoOptionsPropertyMock';
            default:
                return '';
        }
    }

    private ControlToType(control: IData['Inputs'][any]) {
        switch (control) {
            case 'DateAndTime.DateOnly':
                return 'Date';
            case 'Multiple':
            case 'SingleLine.Text':
                return 'string';
            case 'TwoOptions':
                return 'boolean';
            default:
                return '';
        }
    }

    async Render(data: IData) {
        const buffer = await readFile(require.resolve('./renderGenerator.template.hbs'));
        const template = Handlebars.compile(buffer.toString());


        const dts = template({ data });

        return dts;
    }
}