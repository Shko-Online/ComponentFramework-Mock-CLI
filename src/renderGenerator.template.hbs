/* 
    This file was generated with @shko.online/componentframework-mock-cli (pcf-sb)
    Do not manually update this file as it gets overwritten by pcf-sb based on the Manifest configuration.
*/

import type { Meta, StoryObj } from '@storybook/html';

import type { IInputs, IOutputs } from '{{data.ComponentFolder}}/generated/ManifestTypes';

import { useArgs, useEffect } from "@storybook/preview-api";

import { {{data.MockGenerator}}, {{GetPropertyMocks data.Inputs}} } from '@shko.online/componentframework-mock';

import { {{data.Component}} as Component } from '{{data.ComponentFolder}}/{{data.ComponentIndex}}';
 
{{#if (isVirtualComponent data.MockGenerator)}}
import ReactDOM from 'react-dom';
{{/if}}

{{#each data.Styles}}
    import from '{{data.ComponentFolder}}/{{this}}';
{{/each}}

export type StoryArgs {
  isDisabled: boolean;
  isVisible: boolean;
{{#each data.Inputs}}
  {{this.name}}: {{ControlToType this}};
{{/each}}
}

const renderGenerator = (
  onInitialize?: (
    mockGenerator: {{data.MockGenerator}}<IInputs, IOutputs>, 
    container, 
    args: StoryArgs, 
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void, 
  onUpdate?: (
    mockGenerator: {{data.MockGenerator}}<IInputs, IOutputs>, 
    container, 
    args: StoryArgs, 
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void {
  let container: HTMLDivElement | null;
  let mockGenerator: {{data.MockGenerator}}<IInputs, IOutputs>;

  return function () {
    const [args, updateArgs] = useArgs<StoryArgs>();
    // Fires on unload story
    useEffect(
      () => () => {
        container = null;
        mockGenerator.control.destroy();
      },
      []
    );
    // init logic: 
    if (!container) {
      container = document.createElement('div');
      // initialize ComponentFramework Mock Generator
      mockGenerator = new {{data.MockGenerator}}(
        Component,
        {
          {{#each data.Inputs}}
            {{this.name}}: {{ControlToPropertyMock this}}            
          {{/each}}
        },
        {{#if (isVirtualComponent data.MockGenerator)}}
        {{else}}
        container,
        {{/if}}
      );

      onInitialize?.(mockGenerator, container, args, updateArgs);

      // Initialize default state of the parameters
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context._SetCanvasItems({
         {{#each data.Inputs}}
            {{this.name}}: args.{{this.name}},           
         {{/each}}
      });

      // Execute the component's init method      
      mockGenerator.ExecuteInit();
    }

    // Update view
    if (mockGenerator) {
      onUpdate?.(mockGenerator, container, args, updateArgs);

      // Update the component's parameters based on storybook's arguments
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.UpdateValues({
         {{#each data.Inputs }}
            {{this.name}}: args.{{this.name}},       
         {{/each}}
      });
      
      // Execute the component's updateView method 
    {{#if (isVirtualComponent data.MockGenerator)}}
      ReactDOM.render(mockGenerator.ExecuteUpdateView(), container);     
    {{else}}
      mockGenerator.ExecuteUpdateView();
    {{/if}}
    }

    // returning the rendered html component after this update
    return container;
  };
};

export default renderGenerator;
