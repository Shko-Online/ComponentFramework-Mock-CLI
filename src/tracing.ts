import chalk from "chalk";

export const info = (message: string) => {
  console.info(
    chalk.white.bgBlue.bold("[ComponentFramework-Mock-CLI | INFO]") +
      " " +
      message
  );
};

export const error = (message?: string, ...optionalParams: any[]) => {
  console.error(
    chalk.white.bgRed.bold("[ComponentFramework-Mock-CLI | ERROR]") +
      " " +
      message,
    ...optionalParams
  );
};
