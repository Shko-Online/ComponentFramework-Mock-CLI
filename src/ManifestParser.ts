import libxmljs, { XMLElement } from "libxmljs";
import { info } from "./tracing";
import { IData, IProperty } from "ShkoOnline.types";

export class ManifestParser {
  private VERBOSE: boolean;
  constructor(VERBOSE: boolean) {
    this.VERBOSE = VERBOSE;
  }
  async Parse(manifest: Buffer) {
    const xmlDoc = await libxmljs.parseXmlAsync(manifest);
    const control = xmlDoc.find("/manifest/control")[0] as XMLElement;
    const result = {
      Component: control.getAttribute("constructor")?.value(),
      ComponentIndex: (
        xmlDoc.find(
          "/manifest/control/resources/code[@order='1']"
        )[0] as XMLElement
      )
        .getAttribute("path")
        ?.value() as string,
      MockGenerator:
        control.getAttribute("control-type")?.value() === "virtual"
          ? "ComponentFrameworkMockGeneratorReact"
          : "ComponentFrameworkMockGenerator",
      Inputs: xmlDoc.find("/manifest/control/property").map((p) => {
        const type = (p as XMLElement)
          .getAttribute("of-type")
          ?.value() as IProperty["type"];
        const name = (p as XMLElement).getAttribute("name")?.value() as string;
        const usage = (p as XMLElement)
          .getAttribute("usage")
          ?.value() as IProperty["usage"];
        if (type === "Enum") {
          const enumValues = p.find("value").map((v) => ({
            name: (v as XMLElement).getAttribute("name")?.value() as string,
            value: (v as XMLElement).text(),
          }));
          return {
            type,
            name,
            usage,
            enumValues,
          };
        }
        return {
          type,
          name,
          usage,
        };
      }),
    } as IData;

    return result;
  }
}
