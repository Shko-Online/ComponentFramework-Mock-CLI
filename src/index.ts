#!/usr/bin/env node

import { ProgramOptions } from "./ProgramOptions";
import { Option, program } from "commander";
import { error, info } from "./tracing";
import path from "path";
import fs, { constants } from "fs/promises";
import { ManifestParser } from "./ManifestParser";
import Renderer from "./Renderer";

program.version(require("../package.json").version).name("pcf-cli");

program.requiredOption(
  "-c, --component <component>",
  "path to the PCF component"
);
program.addOption(
  new Option(
    "-o, --outputFolder <outputFolder>",
    "path to the desired output folder"
  ).default(".")
);
program.addOption(
  new Option(
    "-r, --renderGenerator <renderGenerator>",
    "override generated file name"
  ).default("{Component}.renderGenerator.ts")
);
program.addOption(
  new Option("-l, --log-level <logLevel>", "specify log level")
    .choices(["none", "error", "verbose"])
    .default("error")
);

const Run = async () => {
  await program.parseAsync();
  const options = program.opts() as ProgramOptions;
  info(JSON.stringify(options));
  const component = path.resolve(options.component);
  try {
    await fs.access(component, constants.R_OK);
  } catch (e) {
    if (options.logLevel !== "none") error("Could not read file.", e);
    process.exit(-1);
  }

  try {
    const parser = new ManifestParser(options.logLevel === "verbose");
    if (options.logLevel === "verbose") {
      info(
        `Reading component manifest from '${path.relative(".", component)}'`
      );
    }
    const manifestXml = await fs.readFile(component);

    if (options.logLevel === "verbose") {
      info("Parsing manifest");
    }
    const manifest = await parser.Parse(manifestXml);
    const renderer = new Renderer();

    const renderGenerator = path.resolve(
      options.outputFolder,
      options.renderGenerator === "{Component}.renderGenerator.ts"
        ? `${manifest.Component}.renderGenerator.ts`
        : options.renderGenerator
    );

    manifest.ComponentFolder = path
      .relative(path.dirname(renderGenerator), path.dirname(component))
      .split(path.sep)
      .join(path.posix.sep);

    await fs.writeFile(renderGenerator, await renderer.Render(manifest));
  } catch (e) {
    if (options.logLevel !== "none")
      error("Could not create renderGenerator.", e);
    process.exit(-1);
  }
};

Run();
