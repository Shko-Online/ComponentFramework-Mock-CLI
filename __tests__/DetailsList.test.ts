import { describe, expect, test } from "@jest/globals";
import { ManifestParser } from "../src/ManifestParser";
import fs from "fs/promises";
import Renderer from "../src/Renderer";

describe("DetailsList", () => {
  test("can generate renderGenerator", async () => {
    const manifestXml = await fs.readFile(
      require.resolve("./DetailsList/ControlManifest.Input.xml")
    );
    const parser = new ManifestParser(true);
    const manifest = await parser.Parse(manifestXml);
    const renderer = new Renderer();
    expect(await renderer.Render(manifest)).toMatchSnapshot();
  });
});
