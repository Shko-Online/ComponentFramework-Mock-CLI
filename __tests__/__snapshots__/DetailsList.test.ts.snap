// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsList can generate renderGenerator 1`] = `
"/* 
    This file was generated with @shko.online/componentframework-mock-cli (pcf-sb)
    Do not manually update this file as it gets overwritten by pcf-sb based on the Manifest configuration.
*/

import type { Meta, StoryObj } from '@storybook/html';

import type { IInputs, IOutputs } from '/generated/ManifestTypes';

import { useArgs, useEffect } from "@storybook/preview-api";

import { ComponentFrameworkMockGeneratorReact, EnumPropertyMock, TwoOptionsPropertyMock, WholeNumberPropertyMock, StringPropertyMock } from '@shko.online/componentframework-mock';

import { FluentDetailsList as Component } from '/index.ts';
 
import ReactDOM from 'react-dom';

import '/css/DetailsList.css';

export type StoryArgs = {
  isDisabled: boolean;
  isVisible: boolean;
  SelectionType: "0" | "1" | "2";
  SelectRowsOnFocus: boolean;
  PageSize: number;
  LargeDatasetPaging: boolean;
  PageNumber: number;
  HasNextPage: boolean;
  HasPreviousPage: boolean;
  TotalRecords: number;
  TotalPages: number;
  CurrentSortColumn: string;
  CurrentSortDirection: "0" | "1";
  AccessibilityLabel: string;
  RaiseOnRowSelectionChangeEvent: boolean;
  InputEvent: string;
  EventName: string;
  EventColumn: string;
  EventRowKey: string;
  SortEventColumn: string;
  SortEventDirection: "0" | "1";
  Theme: string;
  Compact: boolean;
  HeaderVisible: boolean;
  AlternateRowColor: string;
  SelectionAlwaysVisible: boolean;
};

const renderGenerator = (
  onInitialize?: (
    mockGenerator: ComponentFrameworkMockGeneratorReact<IInputs, IOutputs>,
    container: HTMLDivElement,
    args: StoryArgs,
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void,
  onUpdate?: (
    mockGenerator: ComponentFrameworkMockGeneratorReact<IInputs, IOutputs>,
    container: HTMLDivElement,
    args: StoryArgs,
    updateArgs: (newArgs: Partial<StoryArgs>) => void
  ) => void
) => {
  let container: HTMLDivElement | null;
  let mockGenerator: ComponentFrameworkMockGeneratorReact<IInputs, IOutputs>;

  return function () {
    const [args, updateArgs] = useArgs<StoryArgs>();
    // Fires on unload story
    useEffect(
      () => () => {
        container = null;
        mockGenerator.control.destroy();
      },
      []
    );
    // init logic: 
    if (!container) {
      container = document.createElement('div');
      // initialize ComponentFramework Mock Generator
      mockGenerator = new ComponentFrameworkMockGeneratorReact(
        Component,
        {
            SelectionType: EnumPropertyMock,            
            SelectRowsOnFocus: TwoOptionsPropertyMock,            
            PageSize: WholeNumberPropertyMock,            
            LargeDatasetPaging: TwoOptionsPropertyMock,            
            PageNumber: WholeNumberPropertyMock,            
            HasNextPage: TwoOptionsPropertyMock,            
            HasPreviousPage: TwoOptionsPropertyMock,            
            TotalRecords: WholeNumberPropertyMock,            
            TotalPages: WholeNumberPropertyMock,            
            CurrentSortColumn: StringPropertyMock,            
            CurrentSortDirection: EnumPropertyMock,            
            AccessibilityLabel: StringPropertyMock,            
            RaiseOnRowSelectionChangeEvent: TwoOptionsPropertyMock,            
            InputEvent: StringPropertyMock,            
            EventName: StringPropertyMock,            
            EventColumn: StringPropertyMock,            
            EventRowKey: StringPropertyMock,            
            SortEventColumn: StringPropertyMock,            
            SortEventDirection: EnumPropertyMock,            
            Theme: StringPropertyMock,            
            Compact: TwoOptionsPropertyMock,            
            HeaderVisible: TwoOptionsPropertyMock,            
            AlternateRowColor: StringPropertyMock,            
            SelectionAlwaysVisible: TwoOptionsPropertyMock,            
        },
        {
           PageNumber: 'number', 
           HasNextPage: 'boolean', 
           HasPreviousPage: 'boolean', 
           TotalRecords: 'number', 
           TotalPages: 'number', 
           EventName: 'string', 
           EventColumn: 'string', 
           EventRowKey: 'string', 
           SortEventColumn: 'string', 
           SortEventDirection: 'ERROR!Enum', 
        }
      );

      onInitialize?.(mockGenerator, container, args, updateArgs);

      // Initialize default state of the parameters
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context._SetCanvasItems({
            SelectionType: args.SelectionType,           
            SelectRowsOnFocus: args.SelectRowsOnFocus,           
            PageSize: args.PageSize,           
            LargeDatasetPaging: args.LargeDatasetPaging,           
            PageNumber: args.PageNumber,           
            HasNextPage: args.HasNextPage,           
            HasPreviousPage: args.HasPreviousPage,           
            TotalRecords: args.TotalRecords,           
            TotalPages: args.TotalPages,           
            CurrentSortColumn: args.CurrentSortColumn,           
            CurrentSortDirection: args.CurrentSortDirection,           
            AccessibilityLabel: args.AccessibilityLabel,           
            RaiseOnRowSelectionChangeEvent: args.RaiseOnRowSelectionChangeEvent,           
            InputEvent: args.InputEvent,           
            EventName: args.EventName,           
            EventColumn: args.EventColumn,           
            EventRowKey: args.EventRowKey,           
            SortEventColumn: args.SortEventColumn,           
            SortEventDirection: args.SortEventDirection,           
            Theme: args.Theme,           
            Compact: args.Compact,           
            HeaderVisible: args.HeaderVisible,           
            AlternateRowColor: args.AlternateRowColor,           
            SelectionAlwaysVisible: args.SelectionAlwaysVisible,           
      });

      // Execute the component's init method      
      mockGenerator.ExecuteInit();
    }

    // Update view
    if (mockGenerator) {
      onUpdate?.(mockGenerator, container, args, updateArgs);

      // Update the component's parameters based on storybook's arguments
      mockGenerator.context.mode.isVisible = args.isVisible;
      mockGenerator.context.mode.isControlDisabled = args.isDisabled;
      mockGenerator.UpdateValues({
            SelectionType: args.SelectionType,       
            SelectRowsOnFocus: args.SelectRowsOnFocus,       
            PageSize: args.PageSize,       
            LargeDatasetPaging: args.LargeDatasetPaging,       
            PageNumber: args.PageNumber,       
            HasNextPage: args.HasNextPage,       
            HasPreviousPage: args.HasPreviousPage,       
            TotalRecords: args.TotalRecords,       
            TotalPages: args.TotalPages,       
            CurrentSortColumn: args.CurrentSortColumn,       
            CurrentSortDirection: args.CurrentSortDirection,       
            AccessibilityLabel: args.AccessibilityLabel,       
            RaiseOnRowSelectionChangeEvent: args.RaiseOnRowSelectionChangeEvent,       
            InputEvent: args.InputEvent,       
            EventName: args.EventName,       
            EventColumn: args.EventColumn,       
            EventRowKey: args.EventRowKey,       
            SortEventColumn: args.SortEventColumn,       
            SortEventDirection: args.SortEventDirection,       
            Theme: args.Theme,       
            Compact: args.Compact,       
            HeaderVisible: args.HeaderVisible,       
            AlternateRowColor: args.AlternateRowColor,       
            SelectionAlwaysVisible: args.SelectionAlwaysVisible,       
      });
      
      // Execute the component's updateView method 
      ReactDOM.render(mockGenerator.ExecuteUpdateView(), container);     
    }

    // returning the rendered html component after this update
    return container;
  };
};

export default renderGenerator;
"
`;
